feat(tests): add `range` supports

lowlighter opened the pull request:
May close #263 and close #288

This add support for the `range` attribute that can be passed in a step context.
The following types along their contexted variables are supported:
- `[]interface{}`: an array of values
  - `.index`/`.key`: current iteration index
  - `.value`: current iteration item value
- `map[string]interface{}`: a map of values (each key will be iterated over)
  - `.index`: current iteration index
  - `.key`: current iteration item key
  - `.value`: current iteration item value
- `int`: an integer to perform action `n` times
  - `.index`/`.key`/`.value`: current iteration index
- `string`: a templated string which results in one of the above typing 
  - It can be either inherited from vars file, or interpolated from a previous step result 

Within a range context, `.range` variable will be set to `true` (while it'll be `false` in "regular" contexts)
 
Here's an example (see `tests/ranged.yml` in this PR diff for more examples):
```yaml
- name: range with harcoded array
  steps:
  - type: exec
    range:
      - actual: hello
        expected: hello
        index: 0
      - actual: world
        expected: world
        index: 1
    script: echo "{{.range}} {{.index}} {{.value.actual}}"
    assertions:
    - result.code ShouldEqual 0
    - result.systemout ShouldEqual "true {{.value.index}} {{.value.expected}}"
```
Image below show the content of `Range struct` for each test case:
![image](https://user-images.githubusercontent.com/22963968/143625676-514271e2-5c84-4796-af74-5a713307b130.png)

____

I haven't run `go fmt` yet as it'll enclose the current step processing (L176 to L281) within the `for` which make the diff harder to review. If no further changes are requested, I'll run the formatter and add some documentation in the `README.md` with definitive syntax :slightly_smiling_face:

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+
+	//Load "range" attribute and perform actions depending on its typing
+	var ranged Range
+	json.Unmarshal([]byte(rawStep), &ranged)
```
```suggestion
	if err := json.Unmarshal(rawStep, &ranged); err != nil {
	     return nil, fmt.Errorf("unable to parse range expression: %v", err)
	} 
```

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+			Debug(ctx, "attempting to template \"range\" attribute and parse it again")
+			vars, err := DumpStringPreserveCase(stepVars)
+			if err != nil {
+				break
```
you should log the error

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+			}
+			content, err := interpolate.Do(string(rawStep), vars)
+			if err != nil {
+				break
```
you should log the error

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+			//Try parsing data
+			err = json.Unmarshal([]byte(content), &ranged)
+			if err != nil {
+				break
```
you should log the error

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+			switch ranged.RawContent.(type) {
+			case string:
+				rawString = ranged.RawContent.(string)
+				json.Unmarshal([]byte("{\"range\":"+rawString+"}"), &ranged)
```
you should check the error

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+		}
+
+		// Try parsing already templated data
+		err := json.Unmarshal([]byte("{\"range\":"+rawString+"}"), &ranged)
+
+		// ... or fallback
+		if err != nil {
```
```suggestion
		err := json.Unmarshal([]byte("{\"range\":"+rawString+"}"), &ranged)
		// ... or fallback
		if err != nil {
```

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+		}
+
+	//Map-like data
+	case map[string]interface{}:
```
```suggestion
	case map[string]interface{},map[interface{}]interface{}:
```

lowlighter replied:
Is the aim of using `map[interface{}]interface{}` for complex mapping ?

Like: 
```yaml
range:
      ? {key: foo}
      : {actual: hello, expected: hello}
      ? {key: bar}
      : {actual: world, expected: world}	
```
If that's the case I don't think it's supported from what I've seen [go.dev/play/p/4yC3E54dLGy](https://go.dev/play/p/4yC3E54dLGy) though I may have missed something

fsamin replied:
tbh i don't remember when/how it's possible to get `map[interface{}]interface{}` out of an unmarshaler, you can skip this comment, if you prefer

lowlighter replied:
ok I'll let as it since I don't have a test case for it 
I don't think it's possible to make dynamic type assertions, so it would actually require to create a new case to handle specifically `ranged.RawContent.(map[interface{}]interface{})`

fsamin commented on process_testcase.go:
```diff
@@ -270,7 +287,100 @@ func (v *Venom) runTestSteps(ctx context.Context, tc *TestCase) {
+
+	//Unsupported data format
+	default:
+		fmt.Println(ranged.RawContent)
```
why fmt.Println ?

lowlighter replied:
was a leftover from debugging, good catch

fsamin reviewed (changes requested):
Thanks for the PR.
Here are some feedback.

Can you also add a new section in the README ?

lowlighter commented:
- Applied feedback changes and rebased to fix conflict with #452
- Updated the `README.md` to add section about `range` expression
- Also applied `go fmt` (which makes the diff harder to read because it changes the indents, can revert it temporarly for review if you want)

fsamin commented on README.md:
```diff
@@ -508,6 +509,46 @@ testcases:
+    range:
+      - actual: hello
+        expected: hello
+        index: 0
```
avoid this field in the map example, it can be confusing to have `value.index` and `.index`

lowlighter replied:
it's mostly for testing that `.index` is effectively templated and set to correct value, but from an user pov it makes no sense
maybe just renaming `index` to `testIndex` would be better ?

fsamin replied:
it's fine to keep it in our tests files. but avoid it in the README.md

lowlighter replied:
ok I removed it from `README.md`

fsamin commented on README.md:
```diff
@@ -508,6 +509,46 @@ testcases:
+- A templated string which results in one of the above typing (`string`)
+  - It can be either inherited from vars file, or interpolated from a previous step result
+
+Within a range context, `.range` contexted variable will be set to `true` (while it'll be `false` in regular contexts)
```
in fact i'm not sure the `{{.range}}` is still useful ? how do you think you will use it ?

lowlighter replied:
same as above, it's mostly for unit test to check in which context the step is running

fsamin replied:
i guess the other tests can assert the whole feature is working. i prefer to remove it if it's not strictly required.

lowlighter replied:
ok I removed all references to `{{.range}}` and also removed the go code that was setting it
